File: bindings/swift/Sources/Zag/ZagProcess.swift
import Foundation

/// Subprocess helpers for invoking the zag CLI.
public enum ZagProcess {

    /// Default zag binary path from `ZAG_BIN` env or `"zag"`.
    public static var defaultBin: String {
        ProcessInfo.processInfo.environment["ZAG_BIN"] ?? "zag"
    }

    // MARK: - exec

    /// Run zag and return parsed `AgentOutput`.
    public static func exec(bin: String, args: [String]) async throws -> AgentOutput {
        let (stdout, stderr, exitCode) = try await collectOutput(bin: bin, args: args)

        if exitCode != 0 {
            let detail = stderr.isEmpty ? stdout : stderr
            throw ZagError(
                message: "zag exited with code \(exitCode): \(detail)",
                exitCode: Int(exitCode),
                stderr: stderr)
        }

        guard let data = stdout.data(using: .utf8) else {
            throw ZagError(
                message: "Failed to read zag output as UTF-8",
                exitCode: Int(exitCode),
                stderr: stderr)
        }

        do {
            return try JSONDecoder.zag.decode(AgentOutput.self, from: data)
        } catch {
            let preview = String(stdout.prefix(200))
            throw ZagError(
                message: "Failed to parse zag JSON output: \(preview)",
                exitCode: Int(exitCode),
                stderr: stderr)
        }
    }

    // MARK: - stream

    /// Run zag in streaming mode, yielding `Event` objects from NDJSON.
    public static func stream(bin: String, args: [String]) -> AsyncThrowingStream<Event, Error> {
        let (executableURL, fullArgs) = resolveCommand(bin: bin, args: args)

        return AsyncThrowingStream { continuation in
            let task = Task {
                let process = Process()
                process.executableURL = executableURL
                process.arguments = fullArgs

                let stdoutPipe = Pipe()
                let stderrPipe = Pipe()
                process.standardOutput = stdoutPipe
                process.standardError = stderrPipe
                process.standardInput = FileHandle.nullDevice

                do {
                    try process.run()
                } catch {
                    continuation.finish(throwing: ZagError(
                        message: "Failed to start '\(bin)': \(error)",
                        exitCode: nil, stderr: ""))
                    return
                }

                let handle = stdoutPipe.fileHandleForReading
                for try await line in handle.bytes.lines {
                    let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                    if trimmed.isEmpty { continue }
                    guard let data = trimmed.data(using: .utf8) else { continue }
                    do {
                        let event = try JSONDecoder.zag.decode(Event.self, from: data)
                        continuation.yield(event)
                    } catch {
                        continue
                    }
                }

                process.waitUntilExit()

                if process.terminationStatus != 0 {
                    let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()
                    let stderr = String(data: stderrData, encoding: .utf8) ?? ""
                    continuation.finish(throwing: ZagError(
                        message: "zag exited with code \(process.terminationStatus)",
                        exitCode: Int(process.terminationStatus),
                        stderr: stderr))
                } else {
                    continuation.finish()
                }
            }

            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }

    // MARK: - startStreamingProcess

    /// Start a streaming process with piped stdin and stdout.
    public static func startStreamingProcess(bin: String, args: [String]) throws -> StreamingSession {
        let (executableURL, fullArgs) = resolveCommand(bin: bin, args: args)

        let process = Process()
        process.executableURL = executableURL
        process.arguments = fullArgs

        let stdinPipe = Pipe()
        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        process.standardInput = stdinPipe
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        do {
            try process.run()
        } catch {
            throw ZagError(
                message: "Failed to start '\(bin)': \(error)",
                exitCode: nil, stderr: "")
        }

        return StreamingSession(
            process: process,
            stdinPipe: stdinPipe,
            stdoutPipe: stdoutPipe,
            stderrPipe: stderrPipe)
    }

    // MARK: - run (interactive)

    /// Run zag interactively with inherited stdio.
    public static func runInteractive(bin: String, args: [String]) async throws {
        let (executableURL, fullArgs) = resolveCommand(bin: bin, args: args)

        let process = Process()
        process.executableURL = executableURL
        process.arguments = fullArgs

        do {
            try process.run()
        } catch {
            throw ZagError(
                message: "Failed to start '\(bin)': \(error)",
                exitCode: nil, stderr: "")
        }

        process.waitUntilExit()

        if process.terminationStatus != 0 {
            throw ZagError(
                message: "zag exited with code \(process.terminationStatus)",
                exitCode: Int(process.terminationStatus),
                stderr: "")
        }
    }

    // MARK: - Private helpers

    private static func collectOutput(
        bin: String, args: [String]
    ) async throws -> (stdout: String, stderr: String, exitCode: Int32) {
        let (executableURL, fullArgs) = resolveCommand(bin: bin, args: args)

        let process = Process()
        process.executableURL = executableURL
        process.arguments = fullArgs

        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe
        process.standardInput = FileHandle.nullDevice

        do {
            try process.run()
        } catch {
            throw ZagError(
                message: "Failed to start '\(bin)': \(error)",
                exitCode: nil, stderr: "")
        }

        let stdoutData = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
        let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()

        process.waitUntilExit()

        let stdout = String(data: stdoutData, encoding: .utf8) ?? ""
        let stderr = String(data: stderrData, encoding: .utf8) ?? ""

        return (stdout, stderr, process.terminationStatus)
    }

    /// Resolve binary to an executable URL and adjusted arguments.
    /// If `bin` contains a `/`, it's treated as a direct path.
    /// Otherwise, `/usr/bin/env` is used to locate it via PATH.
    private static func resolveCommand(bin: String, args: [String]) -> (URL, [String]) {
        if bin.contains("/") {
            return (URL(fileURLWithPath: bin), args)
        }
        return (URL(fileURLWithPath: "/usr/bin/env"), [bin] + args)
    }
}

// MARK: - StreamingSession

/// A live streaming session with piped stdin and stdout.
public final class StreamingSession {
    private let process: Process
    private let stdinPipe: Pipe
    private let stdoutPipe: Pipe
    private let stderrPipe: Pipe

    init(process: Process, stdinPipe: Pipe, stdoutPipe: Pipe, stderrPipe: Pipe) {
        self.process = process
        self.stdinPipe = stdinPipe
        self.stdoutPipe = stdoutPipe
        self.stderrPipe = stderrPipe
    }

    /// Send a raw NDJSON line to the agent's stdin.
    public func send(_ message: String) throws {
        guard let data = (message + "\n").data(using: .utf8) else { return }
        stdinPipe.fileHandleForWriting.write(data)
    }

    /// Send a user message to the agent.
    public func sendUserMessage(_ content: String) throws {
        let payload: [String: String] = ["type": "user_message", "content": content]
        let data = try JSONSerialization.data(withJSONObject: payload)
        guard let json = String(data: data, encoding: .utf8) else { return }
        try send(json)
    }

    /// Close stdin to signal no more input.
    public func closeInput() {
        stdinPipe.fileHandleForWriting.closeFile()
    }

    /// Async stream of parsed `Event` objects from stdout.
    public var events: AsyncThrowingStream<Event, Error> {
        let handle = stdoutPipe.fileHandleForReading
        return AsyncThrowingStream { continuation in
            Task {
                for try await line in handle.bytes.lines {
                    let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                    if trimmed.isEmpty { continue }
                    guard let data = trimmed.data(using: .utf8) else { continue }
                    do {
                        let event = try JSONDecoder.zag.decode(Event.self, from: data)
                        continuation.yield(event)
                    } catch {
                        continue
                    }
                }
                continuation.finish()
            }
        }
    }

    /// Wait for the process to exit. Throws `ZagError` on non-zero exit.
    public func wait() async throws {
        closeInput()
        process.waitUntilExit()

        if process.terminationStatus != 0 {
            let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()
            let stderr = String(data: stderrData, encoding: .utf8) ?? ""
            throw ZagError(
                message: "zag exited with code \(process.terminationStatus)",
                exitCode: Int(process.terminationStatus),
                stderr: stderr)
        }
    }
}

File: bindings/java/src/main/java/io/zag/ZagProcess.java
package io.zag;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

/** Subprocess helpers for invoking the zag CLI. */
public final class ZagProcess {

    static final ObjectMapper MAPPER =
            new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private ZagProcess() {}

    /** Get the default zag binary path from ZAG_BIN env or "zag". */
    public static String defaultBin() {
        String env = System.getenv("ZAG_BIN");
        return env != null ? env : "zag";
    }

    /** Run zag and return parsed AgentOutput. */
    public static AgentOutput exec(String bin, List<String> args) throws ZagException {
        List<String> command = buildCommand(bin, args);
        try {
            ProcessBuilder pb = new ProcessBuilder(command);
            pb.redirectErrorStream(false);
            Process process = pb.start();

            String stdout =
                    new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
            String stderr =
                    new String(process.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                String msg = stderr.isEmpty() ? stdout : stderr;
                throw new ZagException(
                        "zag exited with code " + exitCode + ": " + msg, exitCode, stderr);
            }

            return MAPPER.readValue(stdout, AgentOutput.class);
        } catch (ZagException e) {
            throw e;
        } catch (IOException | InterruptedException e) {
            throw new ZagException("Failed to run zag: " + e.getMessage(), null, "");
        }
    }

    /** Run zag in streaming mode, returning an iterable of Event objects from NDJSON. */
    public static Iterable<Event> stream(String bin, List<String> args) throws ZagException {
        List<String> command = buildCommand(bin, args);
        try {
            ProcessBuilder pb = new ProcessBuilder(command);
            pb.redirectErrorStream(false);
            Process process = pb.start();

            BufferedReader reader =
                    new BufferedReader(
                            new InputStreamReader(
                                    process.getInputStream(), StandardCharsets.UTF_8));

            return () ->
                    new Iterator<>() {
                        private Event next;
                        private boolean done;

                        @Override
                        public boolean hasNext() {
                            if (done) return false;
                            if (next != null) return true;
                            try {
                                while (true) {
                                    String line = reader.readLine();
                                    if (line == null) {
                                        done = true;
                                        try {
                                            int exitCode = process.waitFor();
                                            if (exitCode != 0) {
                                                String stderr =
                                                        new String(
                                                                process.getErrorStream()
                                                                        .readAllBytes(),
                                                                StandardCharsets.UTF_8);
                                                throw new RuntimeException(
                                                        new ZagException(
                                                                "zag exited with code " + exitCode,
                                                                exitCode,
                                                                stderr));
                                            }
                                        } catch (InterruptedException e) {
                                            Thread.currentThread().interrupt();
                                        }
                                        return false;
                                    }
                                    String trimmed = line.trim();
                                    if (trimmed.isEmpty()) continue;
                                    try {
                                        next = MAPPER.readValue(trimmed, Event.class);
                                        return true;
                                    } catch (Exception e) {
                                        // Skip unparseable lines
                                    }
                                }
                            } catch (IOException e) {
                                done = true;
                                return false;
                            }
                        }

                        @Override
                        public Event next() {
                            if (!hasNext()) throw new NoSuchElementException();
                            Event result = next;
                            next = null;
                            return result;
                        }
                    };
        } catch (IOException e) {
            throw new ZagException("Failed to start zag: " + e.getMessage(), null, "");
        }
    }

    /** Start a streaming process with piped stdin and stdout. */
    public static StreamingSession startStreamingProcess(String bin, List<String> args)
            throws ZagException {
        List<String> command = buildCommand(bin, args);
        try {
            ProcessBuilder pb = new ProcessBuilder(command);
            pb.redirectErrorStream(false);
            Process process = pb.start();
            return new StreamingSession(process);
        } catch (IOException e) {
            throw new ZagException("Failed to start zag: " + e.getMessage(), null, "");
        }
    }

    /** Run zag interactively with inherited stdio. */
    public static void run(String bin, List<String> args) throws ZagException {
        List<String> command = buildCommand(bin, args);
        try {
            ProcessBuilder pb = new ProcessBuilder(command);
            pb.inheritIO();
            Process process = pb.start();
            int exitCode = process.waitFor();
            if (exitCode != 0) {
                throw new ZagException(
                        "zag exited with code " + exitCode, exitCode, "");
            }
        } catch (ZagException e) {
            throw e;
        } catch (IOException | InterruptedException e) {
            throw new ZagException("Failed to run zag: " + e.getMessage(), null, "");
        }
    }

    private static List<String> buildCommand(String bin, List<String> args) {
        List<String> command = new ArrayList<>(args.size() + 1);
        command.add(bin);
        command.addAll(args);
        return command;
    }
}
